File: hooks/lint-check.py
#!/usr/bin/env python3
"""
PostToolUse hook: Run appropriate linter after file edits.

Supports:
- Python (marimo): marimo check
- Python (regular): ruff check
- R: lintr
- Stata: stata-linter
- SAS: sasjs lint
- Markdown: smartquotes --check

Non-blocking: reports linter output as messages.
Silently skips if linter not installed.
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path
from typing import Callable

PLUGIN_ROOT = Path(__file__).parent.parent


def run_command(cmd: list[str], timeout: int = 30) -> tuple[int, str, str]:
    """Run command and return (returncode, stdout, stderr)."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        return result.returncode, result.stdout, result.stderr
    except FileNotFoundError:
        return -1, "", "command not found"
    except subprocess.TimeoutExpired:
        return -2, "", "timeout"
    except Exception as e:
        return -3, "", str(e)


def is_marimo_notebook(file_path: str) -> bool:
    """Check if Python file is a marimo notebook."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read(2000)  # Check first 2000 chars
            return 'import marimo' in content or '@app.cell' in content
    except Exception:
        return False


def check_python(file_path: str) -> str | None:
    """Run marimo check or ruff on Python file."""
    if is_marimo_notebook(file_path):
        # marimo check
        code, stdout, stderr = run_command(['marimo', 'check', file_path])
        if code == -1:
            return None  # marimo not installed
        if code != 0:
            output = (stdout + stderr).strip()
            return f"marimo check:\n{output}" if output else None
    else:
        # ruff check
        code, stdout, stderr = run_command(['ruff', 'check', '--no-fix', file_path])
        if code == -1:
            return None  # ruff not installed
        if code != 0:
            output = stdout.strip()
            return f"ruff:\n{output}" if output else None
    return None


def check_r(file_path: str) -> str | None:
    """Run lintr on R file."""
    cmd = ['Rscript', '-e', f"cat(lintr::lint('{file_path}'))"]
    code, stdout, stderr = run_command(cmd)
    if code == -1:
        return None  # R/lintr not installed
    output = stdout.strip()
    if output and 'Error' not in stderr:
        return f"lintr:\n{output}"
    return None


def check_stata(file_path: str) -> str | None:
    """Run stata-linter on Stata file."""
    code, stdout, stderr = run_command(['stata-linter', file_path])
    if code == -1:
        return None  # stata-linter not installed
    if code != 0 or stdout.strip():
        output = stdout.strip()
        return f"stata-linter:\n{output}" if output else None
    return None


def check_sas(file_path: str) -> str | None:
    """Run sasjs lint on SAS file."""
    code, stdout, stderr = run_command(['sasjs', 'lint', file_path])
    if code == -1:
        return None  # sasjs not installed
    if code != 0 or stdout.strip():
        output = stdout.strip()
        return f"sasjs lint:\n{output}" if output else None
    return None


def check_markdown(file_path: str) -> str | None:
    """Run smartquotes --check on markdown file."""
    script = PLUGIN_ROOT / 'scripts' / 'smartquotes.py'
    if not script.exists():
        return None
    code, stdout, stderr = run_command(['python3', str(script), file_path, '--check'])
    if code == -1:
        return None  # python3 not available
    # smartquotes exits 0 even when changes needed, check output
    output = stdout.strip()
    if output and 'No changes needed' not in output:
        return f"smartquotes:\n{output}\n\nRun: python3 scripts/smartquotes.py {file_path}"
    return None


def get_linter_for_file(file_path: str) -> Callable | None:
    """Get appropriate linter function for file type."""
    path = Path(file_path)
    suffix = path.suffix.lower()

    linters = {
        '.py': check_python,
        '.r': check_r,
        '.R': check_r,
        '.do': check_stata,
        '.ado': check_stata,
        '.sas': check_sas,
        '.md': check_markdown,
        '.markdown': check_markdown,
    }

    return linters.get(suffix)


def main():
    try:
        hook_input = json.load(sys.stdin)
    except Exception:
        sys.exit(0)

    tool_name = hook_input.get('tool_name', '')
    tool_input = hook_input.get('tool_input', {})

    if tool_name not in ('Edit', 'Write'):
        sys.exit(0)

    file_path = tool_input.get('file_path', '')
    if not file_path:
        sys.exit(0)

    # Get appropriate linter
    linter = get_linter_for_file(file_path)
    if not linter:
        sys.exit(0)

    # Run linter
    output = linter(file_path)
    if not output:
        sys.exit(0)

    # Report issues (non-blocking, adds context for Claude)
    result = {
        "hookSpecificOutput": {
            "hookEventName": "PostToolUse",
            "additionalContext": f"Linter output:\n{output}"
        }
    }
    print(json.dumps(result))
    sys.exit(0)


if __name__ == '__main__':
    main()

File: hooks/lint-check.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook: Run appropriate linter after file edits. TypeScript port of lint-check.py.
 *
 * Supports:
 * - Python (marimo): marimo check
 * - Python (regular): ruff check
 * - TypeScript/JavaScript: eslint (project-local or npx)
 * - R: lintr
 * - Stata: stata-linter
 * - SAS: sasjs lint
 *
 * Non-blocking: reports linter output as messages (PostToolUse additionalContext).
 * Silently skips if linter not installed.
 *
 * PORTING NOTES (behavior preserved exactly, oddities included):
 * - `run_command` collapses "binary missing" to returncode -1, timeout to -2, anything else to -3.
 *   Every checker treats ONLY -1 as "linter not installed" and returns silence; a -3 (e.g. a
 *   non-executable candidate eslint) falls through to the `code != 0` branch instead.
 * - The `'.R'` key in the linters map is dead code — the suffix is lowercased before lookup — and is
 *   kept here for the same reason it exists there: this is a port, not a refactor.
 * - Python's `Path.suffix` returns "" for a dotfile with no other dot ('.bashrc') and for a trailing
 *   dot ('x.'); `pySuffix` reproduces that rule rather than a naive lastIndexOf.
 * - Output is serialized with `pyJson`, not `JSON.stringify`, so the bytes match `json.dumps`.
 */
import { context, readPayload } from "./_gate_common.ts";
import { existsSync, readFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";

type CmdResult = { code: number; stdout: string; stderr: string };

/** Run a command, mapping failures onto Python's sentinel return codes (-1/-2/-3). */
function runCommand(cmd: string[], timeout = 30): CmdResult {
  try {
    const proc = Bun.spawnSync(cmd, {
      stdout: "pipe",
      stderr: "pipe",
      stdin: "ignore",
      timeout: timeout * 1000,
    });
    const dec = new TextDecoder();
    const stdout = dec.decode(proc.stdout ?? new Uint8Array());
    const stderr = dec.decode(proc.stderr ?? new Uint8Array());
    // Bun surfaces a timeout kill as signalCode SIGTERM/SIGKILL with the deadline exceeded;
    // Python raises TimeoutExpired -> -2.
    if (proc.signalCode && (proc.signalCode === "SIGTERM" || proc.signalCode === "SIGKILL")) {
      return { code: -2, stdout: "", stderr: "timeout" };
    }
    return { code: proc.exitCode ?? -3, stdout, stderr };
  } catch (e) {
    const msg = String((e as { message?: string })?.message ?? e);
    const code = (e as { code?: string })?.code;
    // FileNotFoundError -> -1; every other exception -> -3.
    if (code === "ENOENT" || /ENOENT|No such file or directory|Executable not found/i.test(msg)) {
      return { code: -1, stdout: "", stderr: "command not found" };
    }
    return { code: -3, stdout: "", stderr: msg };
  }
}

/** Python's `Path.suffix`: last dot-part of the basename, unless leading or trailing. */
function pySuffix(filePath: string): string {
  const name = basename(filePath.replace(/\/+$/, ""));
  const i = name.lastIndexOf(".");
  return i > 0 && i < name.length - 1 ? name.slice(i) : "";
}

function isMarimoNotebook(filePath: string): boolean {
  try {
    const content = readFileSync(filePath, "utf8").slice(0, 2000);
    return content.includes("import marimo") || content.includes("@app.cell");
  } catch {
    return false;
  }
}

function checkPython(filePath: string): string | null {
  if (isMarimoNotebook(filePath)) {
    const { code, stdout, stderr } = runCommand(["marimo", "check", filePath]);
    if (code === -1) return null; // marimo not installed
    if (code !== 0) {
      const output = (stdout + stderr).trim();
      return output ? `marimo check:\n${output}` : null;
    }
  } else {
    const { code, stdout } = runCommand(["ruff", "check", "--no-fix", filePath]);
    if (code === -1) return null; // ruff not installed
    if (code !== 0) {
      const output = stdout.trim();
      return output ? `ruff:\n${output}` : null;
    }
  }
  return null;
}

function checkR(filePath: string): string | null {
  const cmd = ["Rscript", "-e", `cat(lintr::lint('${filePath}'))`];
  const { code, stdout, stderr } = runCommand(cmd);
  if (code === -1) return null; // R/lintr not installed
  const output = stdout.trim();
  if (output && !stderr.includes("Error")) return `lintr:\n${output}`;
  return null;
}

function checkStata(filePath: string): string | null {
  const { code, stdout } = runCommand(["stata-linter", filePath]);
  if (code === -1) return null; // stata-linter not installed
  if (code !== 0 || stdout.trim()) {
    const output = stdout.trim();
    return output ? `stata-linter:\n${output}` : null;
  }
  return null;
}

function checkSas(filePath: string): string | null {
  const { code, stdout } = runCommand(["sasjs", "lint", filePath]);
  if (code === -1) return null; // sasjs not installed
  if (code !== 0 || stdout.trim()) {
    const output = stdout.trim();
    return output ? `sasjs lint:\n${output}` : null;
  }
  return null;
}

/** Walk up from the file to find node_modules/.bin/eslint (max 10 levels). */
function findLocalEslint(filePath: string): string | null {
  let current = dirname(resolve(filePath));
  for (let i = 0; i < 10; i++) {
    const candidate = join(current, "node_modules", ".bin", "eslint");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return null;
}

function checkJsTs(filePath: string): string | null {
  const localEslint = findLocalEslint(filePath);
  const cmd = localEslint
    ? [localEslint, "--no-fix", filePath]
    : ["npx", "--no-install", "eslint", "--no-fix", filePath];

  const { code, stdout } = runCommand(cmd);
  if (code === -1) return null; // eslint not available
  if (code !== 0) {
    const output = stdout.trim();
    return output ? `eslint:\n${output}` : null;
  }
  return null;
}

const LINTERS: Record<string, (p: string) => string | null> = {
  ".py": checkPython,
  ".r": checkR,
  ".R": checkR, // dead in Python too (suffix is lowercased first) — kept for fidelity
  ".do": checkStata,
  ".ado": checkStata,
  ".sas": checkSas,
  ".ts": checkJsTs,
  ".tsx": checkJsTs,
  ".js": checkJsTs,
  ".jsx": checkJsTs,
  ".mjs": checkJsTs,
  ".mts": checkJsTs,
};

let hookInput: Record<string, unknown>;
try {
  hookInput = await readPayload();
} catch {
  process.exit(0);
}

// Python does `hook_input.get(...)` unguarded: a non-dict payload (`null`, `[]`) raises
// AttributeError -> traceback on stderr, exit 1. Reproduced rather than smoothed over.
if (hookInput === null || typeof hookInput !== "object" || Array.isArray(hookInput)) {
  console.error("AttributeError: object has no attribute 'get'");
  process.exit(1);
}

const toolName = String(hookInput.tool_name ?? "");
const toolInput = (hookInput.tool_input ?? {}) as Record<string, unknown>;

if (toolName !== "Edit" && toolName !== "Write") process.exit(0);

const filePath = String(toolInput.file_path ?? "");
if (!filePath) process.exit(0);

const linter = LINTERS[pySuffix(filePath).toLowerCase()];
if (!linter) process.exit(0);

const output = linter(filePath);
if (!output) process.exit(0);

context("PostToolUse", `Linter output:\n${output}`);
